Validate upon encoding (rather than solely on decoding)

joker314 opened the pull request:
Resolves #20 

In that issue, it was expected that Opera did not support pattern matching. Surprise!!

![Screenshot of validation error](https://user-images.githubusercontent.com/18113170/45602430-176f0a00-ba16-11e8-82e6-88a9a8305f34.png)

However, I believe a different related browser doesn't support it (maybe the mobile version or something). Therefore, I still included a little JavaScript.

In addition, this pull request does some other things
- Allows uppercase schemes to not fail the validation checks
- Refactors the regular expressions to change `http|https` to `https?`. Perhaps this may impact readability?

Here's a screenshot of the fallback validation:

![](https://user-images.githubusercontent.com/18113170/45602504-6cf7e680-ba17-11e8-8b85-fe50783cb02d.png)

I don't believe there's a way the unicode URIs could stop working because of this change, the regular expression is in unicode mode (irrelevant), and the surrogate pairs that would make it up should be matched by `.` maybe.

> (But don't explicitly check if the browser is Opera.)

Absolutely.

> For one, if popup windows/tabs are disabled (and 99% of the time they are), would this still work?

Good point, the answer is very probably no.

towerofnix commented on make.html:
```diff
@@ -21,6 +22,17 @@ <h2>What is redirectatob?</h2>
+        // All modern browsers support the pattern attribute, but just in case:
+        if (r && !/^(https?:\/\/|data:)/i.test(r)) {
+          document.getElementById('error').innerText = 'You can only use http://, https://, and data:'
+          document.querySelector('a').innerText = ''
```
(These lines are missing semicolons, btw.)

joker314 replied:
Yep, I copied `querySelector("a")` because it was used later on in the code, but moving it up does make sense. 👍

towerofnix commented on make.html:
```diff
@@ -21,6 +22,17 @@ <h2>What is redirectatob?</h2>
+
+          return 
+        } else {
+          document.getElementById('error').innerText = ''
```
Since you have a `return` above, you may as well take this line out of the `else {}` block (which you can get rid of altogether).

joker314 replied:
Semicolon style: after the `return`, or directly before `document`? Presumably the former since we decided semicolons were good style in this repo?

towerofnix replied:
After `return`.

towerofnix commented on make.html:
```diff
@@ -21,6 +22,17 @@ <h2>What is redirectatob?</h2>
+          document.getElementById('error').innerText = 'You can only use http://, https://, and data:'
+          document.querySelector('a').innerText = ''
+
+          return 
```
(Boring style nit: there shouldn't be a newline above this.)

joker314 commented on make.html:
```diff
@@ -21,6 +22,17 @@ <h2>What is redirectatob?</h2>
+          document.querySelector('a').innerText = ''
+
+          return 
+	} else {
```
Else is being dropped -- resolving

joker314 commented:
Fixed the above. Added unrelated style change, where a string was surrounded in double quotes. All other JavaScript strings in the file used single quotes. (should this be done in a separate PR?)

towerofnix commented:
>Added unrelated style change, where a string was surrounded in double quotes. All other JavaScript strings in the file used single quotes. (should this be done in a separate PR?)

Nah, it's good here, thanks! I meant to fix that as part of some other PR earlier today anyways :P
